modules/Ecommerce/EcoDiscount/Models/EcoDiscount.php

```
@@ -0,0 +1,202 @@
+    }
+
+    /**
+     * Get the orders that used this discount
+     */
+    public function orders(): HasMany
+    {
+        return $this->hasMany(EcoOrder::class, 'discount_id');
+    }
+
+    /**
+     * Check if discount is currently valid
+     */
+    public function isValid(): bool
+    {
+        if (!$this->is_active) {
+            return false;
+        }
+
+        $now = now();
+        
+        if ($this->start_date && $now->lt($this->start_date)) {
+            return false;
+        }
+
+        if ($this->end_date && $now->gt($this->end_date)) {
+            return false;
+        }
+
+        if ($this->usage_limit && $this->used_count >= $this->usage_limit) {
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Check if discount can be applied to given amount
+     */
+    public function canApplyToAmount(float $amount): bool
+    {
+        if (!$this->isValid()) {
+            return false;
+        }
+
+        if ($this->min_order_amount && $amount < $this->min_order_amount) {
+            return false;
+        }
+
+        return true;
+    }
+
+    /**
+     * Calculate discount amount for given order amount
+     */
+    public function calculateDiscountAmount(float $orderAmount): float
+    {
+        if (!$this->canApplyToAmount($orderAmount)) {
+            return 0;
+        }
+
+        $discountAmount = 0;
+
+        switch ($this->type) {
+            case 'percentage':
+                $discountAmount = ($orderAmount * $this->value) / 100;
+                break;
+            case 'fixed_amount':
+                $discountAmount = $this->value;
+                break;
+        }
+
+        // Apply maximum discount limit if set
+        if ($this->max_discount_amount && $discountAmount > $this->max_discount_amount) {
+            $discountAmount = $this->max_discount_amount;
+        }
+
+        // Ensure discount doesn't exceed order amount
+        if ($discountAmount > $orderAmount) {
```

This check should be moved right after calculating the discount amount to fail fast.